lib/1.4/dml-builtins.dml

```diff
@@ -407,6 +407,52 @@ template post_init {
     param _each_post_init = each post_init in (this);
 }
 
+/**
+### destroy
+
+Provides an abstract method `destroy`, which is called when the
+device is being deleted. Typically used to `delete` any device state that has
+been dynamically allocated using `new`.
+
+The method `destroy` is automatically called on all objects that implement the
+`destroy` template. The order in which `destroy` of objects are called is not
+defined, except that `destroy` of a particular object is guaranteed to be
+called before `destroy` of any of its parent objects.
+In particular, `destroy` of the device object will be called only after all
+other implementations of `destroy`.
+
+The `destroy` template can't be instantiated for `event` objects. This is
+because `event` objects, through the `custom_time_event` or
+`custom_cycle_event` template, may already require defining a method named
+`destroy` whose purpose is different from the one required by the `destroy`
+template. To work around this limitation, you may declare a `group` within the
+`event` object that instantiates the `destroy` template instead.
+
+Usage note: while the device is being deleted, it is not allowed to communicate
+with any other Simics object (except the device's `bank`s, `port`s and
+`subdevice`s). This includes even its own clock; don't attempt to post or
```

Comment:
The exception is questionable -- I'd say it's dubious to SIM_get_interface on your own port objects during shutdown (you will get away with it but not sure if we offer any guarantees). And it is *unsafe* to interact with state in private port objs that `destroy` destroys (because of hierarchical destroy order). For portobj-related state owned by the device struct,  like `saved` variables, it is so obviously safe to interact with it during `destroy` that we don't need to mention it.

Reply:
We could instead mention why, e.g. `with any other simics object. This is because that any other objects may be incapable to interaction due to being simultaneously deleted,`

Actually I don't fully remember the details -- are there device interactions that can be safe during `destroy` if you own both devs and know what you are doing, e.g. calling an interface that only relies on data coallocated with the device struct? or does simics reserve the right to deallocate other objects before destroy has been invoked?

Reply:
> The exception is questionable -- I'd say it's dubious to SIM_get_interface on your own port objects during shutdown (you will get away with it but not sure if we offer any guarantees)
> For portobj-related state owned by the device struct, like `saved` variables, it is so obviously safe to interact with it during destroy that we don't need to mention it.

The latter's exactly why I bring it up, You say it's "obviously safe", but I'm not so sure. I just want to preempt the question of "wait, my banks are other objects. does this mean I can't do anything in my banks in `destroy()`"?

> We could instead mention why

That might be a better way to imply why the interaction's unsafe... unless, again, people may wonder "wait, my _banks_ get deleted at the same time, right? Does that mean I can't do anything with them in `destroy()`"
> are there device interactions that can be safe during destroy if you own both devs and know what you are doing?

Not sure but I can't imagine the Simics API enshrines any as acceptable.

Reply:
> n understanding of the model. They'll know what destroy() of each bank/port/subdevices does and doesn't do, and what is safe to interact with within destroy() of, say, the device object. Or am I misunderstanding you?
I meant that interaction with state that is destroyed is unsafe. The following has a 50% probability of crashing, kind of.
```
group x {
  session int *x;
  method init() { x = new int[3]; }
  method destroy() { delete x; }
}
group y {
  method destroy() {
    log info: "%d", x.x[0];
  }
}

Reply:
Oh yes, certainly. My point is that, say, the following is safe:

```
bank b {
  session int *x;
  ...
}

method init() { x = new int[3]; }
method destroy() { delete x; }
```

Reply:
I think the best way to communicate my point is to clarify what is meant by communicate in this context:
```
Usage note: while the device is being deleted, it is not allowed to communicate
with any other Simics object (whether that is by accessing `connect`ed devices
or by leveraging the Simics API.) This includes even the device's own clock...
```

This makes it rather clear I'm not talking about, say, accessing `.val` of registers.

Reply:
I used this phrasing. Take a look and approve or disapprove.